aptos/token_bridge/sources/contract_upgrade.move

```diff
@@ -0,0 +1,147 @@
+    use aptos_framework::code;
+    use wormhole::deserialize;
+    use wormhole::cursor;
+    use wormhole::vaa;
+    use wormhole::state as core;
+
+    use token_bridge::vaa as token_bridge_vaa;
+    use token_bridge::bridge_state as state;
+
+    /// "TokenBridge" (left padded)
+    const TOKEN_BRIDGE: vector<u8> = x"000000000000000000000000000000000000000000546f6b656e427269646765";
+
+    const E_UPGRADE_UNAUTHORIZED: u64 = 0;
+    const E_UNEXPECTED_HASH: u64 = 1;
+    const E_INVALID_MODULE: u64 = 2;
+    const E_INVALID_ACTION: u64 = 3;
+    const E_INVALID_TARGET: u64 = 4;
+
+    // TODO(csongor): document how this works
+    struct UpgradeAuthorized has key {
+        hash: vector<u8>
+    }
+
+    struct Hash {
+        hash: vector<u8>
+    }
+
+    fun parse_payload(payload: vector<u8>): Hash {
+        let cur = cursor::init(payload);
+        let target_module = deserialize::deserialize_vector(&mut cur, 32);
+
+        assert!(target_module == TOKEN_BRIDGE, E_INVALID_MODULE);
+
+        let action = deserialize::deserialize_u8(&mut cur);
+        assert!(action == 0x02, E_INVALID_ACTION);
+
+        let chain = deserialize::deserialize_u16(&mut cur);
+        assert!(chain == core::get_chain_id(), E_INVALID_TARGET);
+
+        let hash = deserialize::deserialize_vector(&mut cur, 32);
+
+        cursor::destroy_empty(cur);
+
+        Hash { hash }
+    }
+
+    public entry fun submit_vaa(vaa: vector<u8>) acquires UpgradeAuthorized {
+        let vaa = vaa::parse_and_verify(vaa);
+        vaa::assert_governance(&vaa);
+        token_bridge_vaa::replay_protect(&vaa);
+
+        authorize_upgrade(parse_payload(vaa::destroy(vaa)));
+    }
+
+    fun authorize_upgrade(hash: Hash) acquires UpgradeAuthorized {
+        let Hash { hash } = hash;
+        let token_bridge = state::token_bridge_signer();
+        if (exists<UpgradeAuthorized>(@token_bridge)) {
+            // TODO(csongor): here we're dropping the upgrade hash, in case an
+            // upgrade fails for some reason. Should we emit a log or something?
+            let UpgradeAuthorized { hash: _ } = move_from<UpgradeAuthorized>(@token_bridge);
+        };
+        move_to(&token_bridge, UpgradeAuthorized { hash });
+    }
+
+    #[test_only]
+    public fun authorized_hash(): vector<u8> acquires UpgradeAuthorized {
+        let u = borrow_global<UpgradeAuthorized>(@token_bridge);
+        u.hash
+    }
+
+    public entry fun upgrade(
+        metadata_serialized: vector<u8>,
+        code: vector<vector<u8>>
+    ) acquires UpgradeAuthorized {
+        assert!(exists<UpgradeAuthorized>(@token_bridge), E_UPGRADE_UNAUTHORIZED);
+        let UpgradeAuthorized { hash } = move_from<UpgradeAuthorized>(@token_bridge);
+
+        let c = copy code;
+        vector::reverse(&mut c);
+        let a = vector::empty<u8>();
+        while (!vector::is_empty(&c)) vector::append(&mut a, vector::pop_back(&mut c));
+        assert!(aptos_hash::keccak256(a) == hash, E_UNEXPECTED_HASH);
+
```

Comment:
no check on metadata_serialized?

Reply:
good point -- I added a TODO item to explore whether this is sufficient or if we have to add the metadata too